pkg/kubefed2/join.go

```diff
@@ -561,20 +626,54 @@ func createClusterRoleAndBinding(clientset client.Interface, saName, namespace,
 			Name:     roleName,
 		},
 	}
-	_, err = clientset.RbacV1().ClusterRoleBindings().Create(binding)
-	if err != nil {
-		glog.V(2).Infof("Could not create cluster role binding for service account: %s in joining cluster: %s due to: %v",
+	existingBinding, err := clientset.RbacV1().ClusterRoleBindings().Get(binding.Name, metav1.GetOptions{})
+	switch {
+	case err != nil && !apierrors.IsNotFound(err):
+		glog.V(2).Infof("Could not get cluster role binding for service account %s in joining cluster %s due to %v",
 			saName, clusterName, err)
 		return err
+	case err == nil && errorOnExisting:
+		return errors.Errorf("cluster role binding for service account %s in joining cluster %s already exists", saName, clusterName)
+	case err == nil:
+		// The roleRef cannot be updated, therefore if the existing roleRef is different, the existing rolebinding
+		// must be deleted and recreated with the correct roleRef
+		if !reflect.DeepEqual(existingBinding.RoleRef, binding.RoleRef) {
+			err = clientset.RbacV1().ClusterRoleBindings().Delete(existingBinding.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				glog.V(2).Infof("Could not delete existing cluster role binding for service account %s in joining cluster %s due to: %v",
+					saName, clusterName, err)
+				return err
+			}
+			_, err = clientset.RbacV1().ClusterRoleBindings().Create(binding)
+			if err != nil {
+				glog.V(2).Infof("Could not create cluster role binding for service account: %s in joining cluster: %s due to: %v",
+					saName, clusterName, err)
+				return err
+			}
+		} else {
+			existingBinding.Subjects = binding.Subjects
+			_, err := clientset.RbacV1().ClusterRoleBindings().Update(existingBinding)
+			if err != nil {
+				glog.V(2).Infof("Could not update cluster role binding for service account: %s in joining cluster: %s due to: %v",
+					saName, clusterName, err)
+				return err
+			}
+		}
+	default:
+		_, err = clientset.RbacV1().ClusterRoleBindings().Create(binding)
+		if err != nil {
+			glog.V(2).Infof("Could not create cluster role binding for service account: %s in joining cluster: %s due to: %v",
+				saName, clusterName, err)
+			return err
+		}
 	}
```

Comment:
To me, this can be simplified with:
```go
case err == nil:
    err = clientset.RbacV1().ClusterRoleBindings().Delete(existingBinding.Name, &metav1.DeleteOptions{})
    ...
    fallthrough
default:
    _, err = clientset.RbacV1().ClusterRoleBindings().Create(binding)
    ...
```

Reasons:
1. more branches in testing
2. `join` is not frequent operation. update vs. delete&update is not critical.

@marun

Reply:
@csrwng Is there a reason you wanted to update an existing binding if possible instead of always deleting?

Reply:
@marun my thinking was that the most common use case is that if there was an existing rolebinding that it would point to the same roleref, which means an update would be just ok. The delete/recreate in my mind is just to cover an edge case.